community/graphdb-api/src/main/java/org/neo4j/graphdb/IdentifiablePropertyContainer.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.graphdb;
+
+/**
+ * Provides the ability to obtain the unique id of the implementing
+ * {@link PropertyContainer}.
+ */
+public interface IdentifiablePropertyContainer extends PropertyContainer
```

Comment:
I am not a fan of the name of this interface.

In fact, I don't like the name `PropertyContainer` either for the responsibilities that interface has (it has too many responsibilities for that name) - and I was the one who introduced that interface way-back-when.

I've been feeling that `PropertyContainer` should only deal with the ability to _read_ property values - that would make it more widely applicable. The ability to write properties should be in a sub-interface that I've been thinking could be called `Entity` (because both `Node` and `Relationship` are entities of the graph database). In `Entity` I would find it less controversial to introduce a `getId()` method (even though I still think the problem of you needing to know the entity type before the ID being useable remains).
The main reason I/we haven't done anything towards this before is because the embedded java API has been down-prioritised in favour of Cypher.